packages/server-boilerplate/src/micro-service/api/ApiBuilder.ts

```diff
@@ -0,0 +1,85 @@
+/**
+ * Tupaia
+ * Copyright (c) 2017 - 2021 Beyond Essential Systems Pty Ltd
+ */
+
+import express, { Express, Request, Response, NextFunction, RequestHandler } from 'express';
+import cors from 'cors';
+import bodyParser from 'body-parser';
+import errorHandler from 'api-error-handler';
+
+import { Authenticator } from '@tupaia/auth';
+import { ModelRegistry, TupaiaDatabase } from '@tupaia/database';
+
+import { handleWith, handleError } from '../../utils';
+import { buildBasicBearerAuthMiddleware } from '../auth';
+import { TestRoute } from '../../routes';
+import { MatchingRequest, Params, ReqBody, ResBody, Query } from '../../routes/Route';
+
+export class ApiBuilder {
+  private readonly app: Express;
+
+  private readonly models: ModelRegistry;
+
+  constructor(transactingConnection: TupaiaDatabase) {
+    this.models = new ModelRegistry(transactingConnection);
+    this.app = express();
+
+    /**
+     * Add middleware
+     */
+    this.app.use(
+      cors({
+        origin: true,
+        credentials: true, // withCredentials needs to be set for cookies to save @see https://developer.mozilla.org/en-US/docs/Web/API/XMLHttpRequest/withCredentials
+      }),
+    );
+    this.app.use(bodyParser.json({ limit: '50mb' }));
+    this.app.use(errorHandler());
+
+    /**
+     * Add singletons to be attached to req for every route
+     */
+    this.app.use((req: Request, res: Response, next: NextFunction) => {
+      req.models = this.models;
+
+      const context = {}; // context is shared between request and response
+      req.ctx = context;
+      res.ctx = context;
+
+      next();
+    });
+
+    /**
+     * Test Route
+     */
+    this.app.get('/v1/test', handleWith(TestRoute));
+  }
+
+  useBasicBearerAuth(apiName: string) {
+    const authenticator = new Authenticator(this.models);
+    this.app.use(buildBasicBearerAuthMiddleware(apiName, authenticator));
+    return this;
+  }
+
+  use<T extends MatchingRequest<T> = Request>(
+    path: string,
+    middleware: RequestHandler<Params<T>, ResBody<T>, ReqBody<T>, Query<T>>,
```

Comment:
I guess that's what you've mentioned in one of your comments: it's not possible to deduce `T` from the passed in params (in this case, middleware), or rather it's too complicated/verbose?

it's fine to use explicit generics (`use<T>`), just trying to get a better understanding of our options 😃

Reply:
Yes, it ought to be possible. What really makes this whole generics code so complicated is because we are working with generics of generics (eg. A generic route, which in turn takes a generic request as a parameter). It seems that typescript doesn't seem to do any kind of 'deep infer' logic, so you need to pass through the nested generic parameters all the way through... it gets very complicated and typescript complains about how complex the type signatures are getting...
But yeah, if we invest the time there should be a way to do it nicely

Reply:
Sounds good, thanks for the explanation! It's already nice as is 😃